federation/pkg/federation-controller/sync/controller.go

```diff
@@ -273,79 +323,65 @@ func (s *FederationSyncController) reconcileOnClusterChange() {
 	}
 }
 
-func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName) {
+func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName) (reconciliationStatus, error) {
 	if !s.isSynced() {
-		s.deliver(namespacedName, s.clusterAvailableDelay, false)
-		return
+		return statusNotSynced, nil
 	}
 
 	key := namespacedName.String()
 	kind := s.adapter.Kind()
 	cachedObj, exist, err := s.store.GetByKey(key)
 	if err != nil {
-		glog.Errorf("Failed to query main %s store for %v: %v", kind, key, err)
-		s.deliver(namespacedName, 0, true)
-		return
+		return statusError, fmt.Errorf("failed to query main %s store for %v: %v", kind, key, err)
 	}
 
 	if !exist {
 		// Not federated, ignoring.
-		return
+		return statusAllOK, nil
 	}
 
 	// Create a copy before modifying the resource to prevent racing
 	// with other readers.
 	copiedObj, err := api.Scheme.DeepCopy(cachedObj)
 	if err != nil {
-		glog.Errorf("Error in retrieving %s from store: %v", kind, err)
-		s.deliver(namespacedName, 0, true)
-		return
+		return statusError, fmt.Errorf("error in retrieving %s from store: %v", kind, err)
 	}
 	if !s.adapter.IsExpectedType(copiedObj) {
-		glog.Errorf("Object is not the expected type: %v", copiedObj)
-		s.deliver(namespacedName, 0, true)
-		return
+		return statusError, fmt.Errorf("object is not the expected type: %v", copiedObj)
 	}
 	obj := copiedObj.(pkgruntime.Object)
 	meta := s.adapter.ObjectMeta(obj)
 
 	if meta.DeletionTimestamp != nil {
 		if err := s.delete(obj, namespacedName); err != nil {
-			glog.Errorf("Failed to delete %s %s: %v", kind, namespacedName, err)
 			s.eventRecorder.Eventf(obj, api.EventTypeWarning, "DeleteFailed",
 				"%s delete failed: %v", strings.ToTitle(kind), err)
-			s.deliver(namespacedName, 0, true)
+			return statusError, fmt.Errorf("failed to delete %s %s: %v", kind, namespacedName, err)
 		}
-		return
+		return statusAllOK, nil
 	}
 
 	glog.V(3).Infof("Ensuring delete object from underlying clusters finalizer for %s: %s",
 		kind, namespacedName)
 	// Add the required finalizers before creating the resource in underlying clusters.
 	obj, err = s.deletionHelper.EnsureFinalizers(obj)
 	if err != nil {
-		glog.Errorf("Failed to ensure delete object from underlying clusters finalizer in %s %s: %v",
+		return statusError, fmt.Errorf("failed to ensure delete object from underlying clusters finalizer in %s %s: %v",
 			kind, namespacedName, err)
-		s.deliver(namespacedName, 0, false)
```

Comment:
That is correct: I'm inclined to believe that returning true before was a typo, since this is an error condition and diverging like this would have warranted at least a comment as to why.